client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,159 @@
+import { emailManagement } from 'calypso/my-sites/email/paths';
+import { composeAnalytics, withAnalytics } from 'calypso/state/analytics/actions';
+import { addEmailForward } from 'calypso/state/email-forwarding/actions';
+import getEmailForwardingLimit from 'calypso/state/selectors/get-email-forwarding-limit';
+import {
+	addEmailForwardSuccess,
+	getEmailForwards,
+	isAddingEmailForward,
+} from 'calypso/state/selectors/get-email-forwards';
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+		};
+	}
+
+	hasValidEmailForwards() {
+		return ! this.state.emailForwards?.some( ( forward ) => ! forward.isValid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination ) => {
+		withAnalytics(
+			composeAnalytics(
+				this.props.onConfirmEmailForwarding(),
+				this.props.addEmailForward( domainName, mailbox, destination )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { selectedSiteSlug, selectedDomainName } = this.props;
+
+		event.preventDefault();
+
+		if ( this.props.isAddingEmailForward ) {
+			return;
+		}
+
+		this.state.emailForwards?.map( ( forward ) => {
+			const { mailbox, destination } = forward;
+
+			this.addNewEmailForwardWithAnalytics(
+				selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+	};
+
+	onForwardAdd = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
+		} );
+	};
+
+	renderActionsButtons( addMoreButton = false ) {
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				{ addMoreButton && (
+					<Button
+						className="email-forwarding-add-new-compact-list__add-another-forward-button"
+						onClick={ this.onForwardAdd }
+					>
+						<Gridicon icon="plus" />
+						<span>{ translate( 'Add another forward' ) }</span>
+					</Button>
+				) }
+
+				<Button
+					primary
+					onClick={ this.addNewEmailForwardsClick }
+					disabled={ ! this.hasValidEmailForwards() }
+				>
+					{ translate( 'Add' ) }
+				</Button>
+			</div>
+		);
+	}
+
+	onRemoveEmailForward = ( index ) => {
+		const [ emailForwards ] = this.state;
+		emailForwards.splice( index, 1 );
+		this.setState( { emailForwards } );
+	};
+
+	onUpdateEmailForward = ( index, name, value ) => {
+		const array = this.state.emailForwards;
+		array[ index ][ name ] = value;
```

Comment:
Two nits:
1. I would find this code easier to understand if we called the variable `emailForwards` instead of `array`.
2. I don't like that we're using a `const` and then updating a value on the next line. The linter may not allow us to use a `let`, but I'd prefer that, as we are modifying the variable.

Reply:
1. I'll rename :)
2. about const, I used const because we are not really modifying the variable (not reassigning it's value eg. memory address/pointer), but modifying it's content. Anyway I get your point, and it's also weird to do some modification on the next line.